### PR TITLE
add inverse_sqrt lr decay style

### PR DESCRIPTION
**Pull request by NouamaneTazi**
adds support for inverse square decay style
P.S: ignores `min-lr` param for now

To use it just add these flags to your command:
```
--lr-decay-style=inverse_sqrt --lr-warmup-style=linear/constant
```

image

### CODE REVIEW

File: megatron/learning_rates.py [hunk 1]
```diff
@@ -68,6 +68,9 @@ def get_lr(self):
             if self.num_steps == self.warmup_steps and \
                 self.decay_tokens is not None:
                 self.warmup_tokens = self.num_tokens
+            # if self.decay_style == 'inverse_sqrt':
+            #     # use constant warmup for inverse_sqrt
+                # return 1e-2
```
Comment (thomasw21) [inline megatron/learning_rates.py]: Actually you should add a "constant warmup" as in T5. (maybe this could be exposed as another argument that accepts "linear" or "constant"). In the constant case you can just result `self.max_lr`

File: megatron/learning_rates.py [hunk 2]
```diff
@@ -102,6 +105,9 @@ def get_lr(self):
             coeff = (1.0 - decay_ratio)
         elif self.decay_style == 'cosine':
             coeff = 0.5 * (math.cos(math.pi * decay_ratio) + 1.0)
+        elif self.decay_style == 'inverse_sqrt':
+            return self.max_lr * math.sqrt(float(max(self.warmup_steps, 1))) / \
+                math.sqrt(float(max(num_steps_,1)))
```
Comment (thomasw21) [inline megatron/learning_rates.py]: Can you write in math formulas what you want here? I don't think it does the same thing as T5 for example in this case. (it should scale in `rsqrt(num_steps)` instead of `sqrt(warmup_step/num_steps)`). Your implementation seems to correspond to https://cs.github.com/google-research/t5x/blob/730967370cab41b4ce51419a40538fe359e07d7b/t5x/utils.py#L793 If you want the original one you should do the following.

```suggestion
            return self.max_lr / math.sqrt(float(max(num_steps_, 1)))
```

Some things to note:
 -  the `min_lr` is going to trigger before, typically if `self.num_token > self.decay_tokens`. So you'll see a big change in `lr` then. Typically I think you might want to move this code before the `if self.decay_tokens` since I don't think you want to support min_lr (otherwise you need to figure out hot to make you lr schedule continuous.
  - I think you should ask around which decay they use for PaLM for example. I'm guessing that it's `rsqrt_decay` according to the paper, but I'm not sure. Either way probably a good thing to support all of them.
   - I'm guessing your using `warmup_steps` the same as `offsets`

Comment (thomasw21) [reply megatron/learning_rates.py]: Also wondering if you actually need to cast to float before?

Comment (nmntz) [reply megatron/learning_rates.py]: We probably dont need to cast as `math.sqrt` return float either way

File: megatron/learning_rates.py (hunk 1 above)
Comment (nmntz) [reply megatron/learning_rates.py]: Fixed, thanks

File: megatron/arguments.py [hunk 3]
```diff
@@ -621,6 +621,8 @@ def _add_learning_rate_args(parser):
     group.add_argument('--lr-decay-tokens', type=int, default=None,
                        help='number of tokens to decay learning rate over,'
                        ' If not None will override iter/sample-based decay')
+    group.add_argument('--lr-warmup-style', type=str, default='linear',
+                       choices=['constant', 'linear']),
```
Comment (thomasw21) [inline megatron/arguments.py]: You can probably add some help documentation, in order to know what does `constant` warmup for example?

File: megatron/learning_rates.py [hunk 4]
```diff
@@ -63,18 +64,27 @@ def get_lr(self):
...
+            else:
+                raise ValueError('Unknown warmup style: {}'.format(
+                    self.warmup_style))
 
         # If the learning rate is constant, just return the initial value.
         if self.decay_style == 'constant':
             return self.max_lr
 
+        if self.decay_style == 'inverse_sqrt':
+            return self.max_lr / math.sqrt(max(self.num_steps - self.warmup_steps, 1))
```
Comment (thomasw21) [inline megatron/learning_rates.py]:
```suggestion
            return self.max_lr / math.sqrt(max(num_steps_, 1))
```

**Review by thomasw21 — APPROVED** [review]
LGTM! You may want to also add the normalized version as well depending on what was used for PaLM for example.

**Comment by thomasw21** [comment]
Oh I don't have write access to this repo. You'll have to ask someone that does. cc @spyysalo

File: megatron/learning_rates.py (hunk 4 above)
Comment (nmntz) [reply megatron/learning_rates.py]: `num_steps_` is still undefined here

Comment (thomasw21) [reply megatron/learning_rates.py]: Yeah so just move it?

Comment (nmntz) [reply megatron/learning_rates.py]: Fixed

File: megatron/learning_rates.py [hunk 5]
```diff
@@ -64,6 +64,19 @@ def get_lr(self):
...
+            if self.warmup_steps > 0 and self.num_steps <= self.warmup_steps:
+                if self.warmup_style == 'linear':
+                    return self.max_lr * self.num_steps / (self.warmup_steps * self.warmup_steps**0.5)
+                elif self.warmup_style == 'constant':
+                    return self.max_lr / self.warmup_steps**0.5
+                else:
+                    raise ValueError('Unknown warmup style: {}'.format(
+                        self.warmup_style))
+                
+            return self.max_lr / (max(self.num_steps, 1))**0.5
```
Comment (thomasw21) [inline megatron/learning_rates.py]: Actually I think I was wrong, but this should be something like 
```
return self.max_lr / (max(self.num_steps, max(self.warmup_steps, 1)))**0.5
```
But I'd argue that this should still be done after :D

File: megatron/learning_rates.py [hunk 6]
```diff
@@ -63,18 +64,32 @@ def get_lr(self):
...
+                raise ValueError('Unknown warmup style: {}'.format(
+                    self.warmup_style))
 
         # If the learning rate is constant, just return the initial value.
         if self.decay_style == 'constant':
             return self.max_lr
 
+
+        # If linear
+        # In warmup phase: lr = max_lr
```
Comment (thomasw21) [inline megatron/learning_rates.py:88]: That comment seems wrong, this applies in constant warmup

File: megatron/learning_rates.py [hunk 7]
```diff
@@ -63,18 +64,32 @@ def get_lr(self):
...
+        # In decay phase: lr = max_lr * sqrt(warmup_steps) / sqrt(num_steps)
```
Comment (thomasw21) [inline megatron/learning_rates.py:89]: Maybe add some sort of comment that max_lr and warmup_steps are related  if you want to replicate PaLM setup.

**Review by thomasw21 — APPROVED** [review]
Lgtm!

File: megatron/learning_rates.py (hunk 6 above)
Comment (nmntz) [reply megatron/learning_rates.py:88]: Oops

File: megatron/learning_rates.py (hunk 7 above)
Comment (nmntz) [reply megatron/learning_rates.py:89]: I'm afraid that'll complicate the comment. Any idea how to put it simply?

Comment (thomasw21) [reply megatron/learning_rates.py:89]: Hum well I guess it's okay like this. I just want for people to be clear that t5x config are not usable here. Typically t5x would be `base_learning_rate=1, warmup_steps=10000` here it would be `max_lr=1e-2, wamup_steps=10000`
